routing/router.go

```diff
@@ -2293,3 +2293,278 @@ func generateBandwidthHints(sourceNode *channeldb.LightningNode,
+			if inEdge == nil {
+				return nil
+			}
+
+			// Before we can process the edge, we'll need to
+			// fetch the node on the _other_ end of this
+			// channel as we may later need to iterate over
+			// the incoming edges of this node if we explore
+			// it further.
+			chanFromNode, err := edgeInfo.FetchOtherNode(
+				tx, toNode[:],
+			)
+			if err != nil {
+				return err
+			}
+
+			// Continue searching if this channel doesn't
+			// connect with the previous hop.
+			if chanFromNode.PubKeyBytes != fromNode {
+				return nil
+			}
+
+			// Validate whether this channel's policy is satisfied
+			// and obtain the new running amounts if this channel
+			// was to be selected.
+			newAmts, err := r.checkOrUpdateAmt(
+				inEdge, edgeInfo.Capacity, amts, localChan,
+				amt == nil,
+			)
+			if err != nil {
+				log.Debugf("Skipping chan %v: %v",
+					inEdge.ChannelID, err)
+
+				return nil
+			}
+
+			// If we already have a best edge, check whether this
+			// edge is better.
+			bandwidth, _ := bandwidthHints[chanID]
+			if bestEdge != nil {
+				if localChan {
+					// For local channels, better is defined
+					// as having more bandwidth. We try to
+					// maximize the chance that the returned
+					// route succeeds.
+					if bandwidth < bestBandwidth {
+						return nil
+					}
+				} else {
+					// For other channels, better is defined
+					// as lower fees for the amount to send.
+					// Normally all channels between two
+					// nodes should have the same policy,
+					// but in case not we minimize our cost
+					// here. Regular path finding would do
+					// the same.
+					if newAmts.amt > bestAmts.amt {
+						return nil
+					}
+				}
+			}
+
+			// If we get here, the current edge is better. Replace
+			// the best.
+			bestEdge = inEdge
+			bestAmts = &newAmts
+			bestBandwidth = bandwidth
+
+			return nil
+		}
+
+		err := r.cfg.Graph.ForEachNodeChannel(nil, toNode[:], cb)
+		if err != nil {
+			return nil, err
+		}
+
+		// There is no matching channel. Stop building the route here.
+		if bestEdge == nil {
+			return nil, ErrNoChannel{
+				fromNode: fromNode,
+				position: i,
+			}
+		}
+
+		log.Debugf("Select channel %v at position %v", bestEdge.ChannelID, i)
+
+		edges[i] = bestEdge
+		amts = *bestAmts
+	}
+
+	_, height, err := r.cfg.Chain.GetBestBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	// If no amount is specified, use the minimum routable amount.
+	var receiverAmt lnwire.MilliSatoshi
+	if amt != nil {
+		receiverAmt = *amt
+	} else {
+		// We've calculated the minimum amount for the htlc that the
+		// source node hands out. The newRoute call below expects the
+		// amount that must reach the receiver after subtraction of fees
+		// along the way. Iterate over all edges to calculate the
+		// receiver amount.
+		receiverAmt = amts.amt
+		for _, edge := range edges[1:] {
+			receiverAmt -= edge.ComputeFeeFromIncoming(receiverAmt)
+		}
+	}
+
+	// Build and return the final route.
+	return newRoute(
```

Comment:
sanity check that the first hop amt is equal to our `amts.amt`?

Reply:
I think this is covered in the unit test where we assert on `route.TotalAmount`. I am no fan of sanity checks.